README.md
# SocialsRegex

----
Social Regex Account Detection and Extraction for Ruby. Detect and extract URLs of social accounts: throw in URLs, get back URLs of social media profiles by type.

Features:

- detect the platform a url points to (all major platforms supported)
- extract the information contained within the url (without opening the url, of course)
- extract emails and phone numbers from hyperlinks

## Installation

Install the gem and add to the application's Gemfile by executing:

    $ bundle add socials_regex

If bundler is not being used to manage dependencies, install the gem by executing:

    $ gem install socials_regex

### Requirements
This gem requires Ruby 2.6+

## Usage

```ruby
require 'socials_regex'

supported_platforms = SocialsRegex::Platforms.all
# [:PLATFORM_FACEBOOK, :PLATFORM_GITHUB, :PLATFORM_LINKEDIN, :PLATFORM_TWITTER, :PLATFORM_INSTAGRAM, :PLATFORM_YOUTUBE, 
# :PLATFORM_EMAIL, :PLATFORM_HACKER_NEWS, :PLATFORM_MEDIUM, :PLATFORM_PHONE, :PLATFORM_REDDIT,
# :PLATFORM_SKYPE, :PLATFORM_SNAPCHAT, :PLATFORM_STACKEXCHANGE, :PLATFORM_STACKOVERFLOW, :PLATFORM_STACKOVERFLOW, 
# :PLATFORM_TELEGRAM, :PLATFORM_VIMEO, :PLATFORM_XING, :PLATFORM_ANGELLIST, :PLATFORM_CRUNCHBASE, 
# :PLATFORM_STACKEXCHANGE_NETWORK, :PLATFORM_WHATSAPP, :PLATFORM_YELP]


supported_regexes = SocialsRegex::Regexes.all
# [:ANGELLIST_URL_REGEX, :CRUNCHBASE_URL_REGEX, :EMAIL_URL_REGEX, :FACEBOOK_URL_REGEX, :GITHUB_URL_REGEX, :HACKERNEWS_URL_REGEX,
# :INSTAGRAM_URL_REGEX, :LINKEDIN_URL_REGEX, :MEDIUM_URL_REGEX, :PHONE_URL_REGEX, :REDDIT_URL_REGEX, :SKYPE_URL_REGEX, :SNAPCHAT_URL_REGEX,
# :STACKEXCHANGE_URL_REGEX, :STACKEXCHANGE_NETWORK_URL_REGEX, :STACKOVERFLOW_URL_REGEX, :TELEGRAM_URL_REGEX, :TWITTER_URL_REGEX,
# :VIMEO_URL_REGEX, :XING_URL_REGEX, :YOUTUBE_URL_REGEX, :WHATSAPP_URL_REGEX, :YELP_URL_REGEX] 

# get all regex for all regex
platform_regexes = SocialsRegex::Socials::PLATFORMS_REGEX
# example [:yelp, {:company=>/(?:https?:\/\/)?(?:www\.)?yelp\.com\/biz\/(?<company>[A-Za-z0-9_-]+)/}] 

# get regex for specific platforms 
twitter_regex = SocialsRegex::Socials::PLATFORMS_REGEX[:twitter]
# {:status=>/(?:https?:)?\/\/(?:[A-Za-z]+\.)?twitter\.com\/@?(?<username>[A-Za-z0-9_]+)\/status\/(?<tweet_id>[0-9]+)\/?/,
# :user=>/(?:https?:)?\/\/(?:[A-Za-z]+\.)?twitter\.com\/@?(?!home|share|privacy|tos)(?<username>[A-Za-z0-9_]+)\/?/}


# how to extract social data from links or texts
text = 'https://twitter.com/karllorey/status/1259924082067374088' \
             'https://twitter.com/karllorey12/status/12599240820673740883' \
             'http://crunchbase.com/organization/acme-corp jeff@amazon.com mailto:plususer+test@gmail.com' \
             'https://facebook.com/peter.parker https://www.facebook.com/profile.php?id=100004123456789' \
             'https://github.com/talaatmagdyx https://github.com/talaatmagdyx/socials_regex' \
             'https://news.ycombinator.com/item?id=23290375 https://instagram.com/__disco__dude' \
             'https://www.linkedin.com/in/talaatmagdyx/ https://medium.com/does-exist/some-post-123abc'
extract = SocialsRegex::Extraction.new(text: text)
# #<SocialsRegex::Extraction:0x00007f5c51d0c488 @text= "https://twitter.com/karllorey/status/......">

# to extract all links and data 
extract.extract_matches_per_platform
# {:crunchbase=>{:company=>[{:matched=>"http://crunchbase.com/organization/acme-corp", "organization"=>"acme-corp"}]},
# :medium=>{:post=>[{:matched=>"https://medium.com/does-exist/some-post-123abc", "username"=>nil, "publication"=>"does-exist", "slug"=>"some-post", "post_id"=>"123abc"}]},
#   :hackernews=>{:item=>[{:matched=>"https://news.ycombinator.com/item?id=23290375", "item"=>"23290375"}]},
#  :email=>{:email=>[{:matched=>"jeff@amazon.com", "email"=>"jeff@amazon.com"}, {:matched=>"mailto:plususer+test@gmail.comhttps", "email"=>"plususer+test@gmail.comhttps"}]},
#    :instagram=>{:profile=>[{:matched=>"https://instagram.com/__disco__dudehttps", "username"=>"__disco__dudehttps"}]},


# to extract links or data using specific platform like instagram
extract.extract_matches_by_platform(platform: 'instagram') # or use :instagram
# {"instagram"=>{:profile=>[{:matched=>"https://instagram.com/__disco__dudehttps", "username"=>"__disco__dudehttps"}]}}

# to extract links or data using specific regex like twitter status
matches = extract.extract_matches_by_regex(regex: SocialsRegex::Regexes::TWITTER_URL_REGEX[:status])
# [{:matched=>"https://twitter.com/karllorey/status/1259924082067374088", "username"=>"karllorey", "tweet_id"=>"1259924082067374088"},
# {:matched=>"https://twitter.com/karllorey12/status/12599240820673740883", "username"=>"karllorey12", "tweet_id"=>"12599240820673740883"}]


```

# References
- [social-media-profiles-regexs](https://github.com/lorey/social-media-profiles-regexs):
  extract urls of social media profiles with regular expressions

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at [Contributing](https://github.com/talaatmagdyx/socials_regex/blob/master/.github/CONTRIBUTING.md). This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/talaatmagdyx/socials_regex/blob/master/.github/CODE_OF_CONDUCT.md).

## [ChangeLog](./CHANGELOG.md)

## Reporting Bugs / Feature Requests

Please [open an Issue on GitHub](https://github.com/talaatmagdyx/socials_regex/issues) if you have feedback, new feature requests, or want to report a bug. Thank you!

## Pull Request
Please read [Contributing](https://github.com/talaatmagdyx/socials_regex/blob/master/.github/PULL_REQUEST_TEMPLATE.md)

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the SocialsRegex project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/talaatmagdyx/socials_regex/blob/master/.github/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [released]

## [1.0.0] - 2023-07-02

- first release


spec/spec_helper.rb
# frozen_string_literal: true

require 'socials_regex'

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = '.rspec_status'

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/socials_regex_spec.rb
# frozen_string_literal: true

RSpec.describe SocialsRegex do
  it 'has a version number' do
    expect(SocialsRegex::VERSION).not_to be_nil
  end
end


lib/socials_regex/socials.rb
# frozen_string_literal: true

module SocialsRegex
  class Socials
    PLATFORMS_REGEX = {
      "#{SocialsRegex::Platforms::PLATFORM_YELP}": SocialsRegex::Regexes::YELP_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_WHATSAPP}": SocialsRegex::Regexes::WHATSAPP_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_STACKEXCHANGE_NETWORK}":
        SocialsRegex::Regexes::STACKEXCHANGE_NETWORK_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_CRUNCHBASE}": SocialsRegex::Regexes::CRUNCHBASE_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_ANGELLIST}": SocialsRegex::Regexes::ANGELLIST_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_XING}": SocialsRegex::Regexes::XING_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_VIMEO}": SocialsRegex::Regexes::VIMEO_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_TELEGRAM}": SocialsRegex::Regexes::TELEGRAM_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_STACKOVERFLOW}": SocialsRegex::Regexes::STACKOVERFLOW_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_STACKEXCHANGE}": SocialsRegex::Regexes::STACKEXCHANGE_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_SNAPCHAT}": SocialsRegex::Regexes::SNAPCHAT_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_SKYPE}": SocialsRegex::Regexes::SKYPE_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_REDDIT}": SocialsRegex::Regexes::REDDIT_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_PHONE}": SocialsRegex::Regexes::PHONE_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_MEDIUM}": SocialsRegex::Regexes::MEDIUM_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_HACKER_NEWS}": SocialsRegex::Regexes::HACKERNEWS_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_EMAIL}": SocialsRegex::Regexes::EMAIL_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_YOUTUBE}": SocialsRegex::Regexes::YOUTUBE_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_INSTAGRAM}": SocialsRegex::Regexes::INSTAGRAM_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_TWITTER}": SocialsRegex::Regexes::TWITTER_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_LINKEDIN}": SocialsRegex::Regexes::LINKEDIN_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_GITHUB}": SocialsRegex::Regexes::GITHUB_URL_REGEX,
      "#{SocialsRegex::Platforms::PLATFORM_FACEBOOK}": SocialsRegex::Regexes::FACEBOOK_URL_REGEX
    }.freeze

    ERROR_MSG_UNKNOWN_PLATFORM = "Unknown platform, expected one of #{PLATFORMS_REGEX.keys}"
  end

  # Extracted profiles.
  class Extraction
    attr_accessor :text

    def initialize(text:)
      @text = text
    end

    #  Get lists of profiles keyed by platform name.
    # :return: a dictionary with the platform as a key, and a list of the platform's profiles as values.
    def extract_matches_per_platform
      matches = {}
      Socials::PLATFORMS_REGEX.each do |platform, regexes|
        matches.merge!(platform_matches(regexes: regexes, platform: platform))
      end
      matches
    end

    # Find all matches for a specific regex.
    # :param regex: platform to search for.
    # :return: list of matches.
    def extract_matches_by_regex(regex:)
      matches(regex: regex)
    end

    # Find all matches for a specific platform.
    # :param platform: platform to search for.
    # :return: hash of list of matches.
    def extract_matches_by_platform(platform:)
      regexes = Socials::PLATFORMS_REGEX[platform.to_sym]
      raise Socials::ERROR_MSG_UNKNOWN_PLATFORM unless regexes

      platform_matches(regexes: regexes, platform: platform)
    end

    private

    # Find all matches for a specific platform.
    # :param regexes:
    # :params platform:
    # :return: hash of list of matches.
    def platform_matches(regexes:, platform:)
      matches = {}
      regexes.each do |key, regex|
        matched = matches(regex: regex)
        next if matched.empty?

        matches[platform] ||= {}
        matches[platform][key] = matched
      end
      matches
    end

    def matches(regex:)
      reformat_matches(matches: text.to_enum(:scan, regex).map { Regexp.last_match })
    end

    def reformat_matches(matches:)
      reformat = []
      matches.each do |match|
        reformat << { matched: match[0] }.merge!(match.named_captures)
      end
      reformat
    end
  end
end


lib/socials_regex/platforms.rb
# frozen_string_literal: true

module SocialsRegex
  # all supported platforms
  class Platforms
    PLATFORM_FACEBOOK = 'facebook'
    PLATFORM_GITHUB = 'github'
    PLATFORM_LINKEDIN = 'linkedin'
    PLATFORM_TWITTER = 'twitter'
    PLATFORM_INSTAGRAM = 'instagram'
    PLATFORM_YOUTUBE = 'youtube'
    PLATFORM_EMAIL = 'email'
    PLATFORM_HACKER_NEWS = 'hackernews'
    PLATFORM_MEDIUM = 'medium'
    PLATFORM_PHONE = 'phone'
    PLATFORM_REDDIT = 'reddit'
    PLATFORM_SKYPE = 'skype'
    PLATFORM_SNAPCHAT = 'snapchat'
    PLATFORM_STACKEXCHANGE = 'stackexchange'
    PLATFORM_STACKOVERFLOW = 'stackoverflow'
    PLATFORM_TELEGRAM = 'telegram'
    PLATFORM_VIMEO = 'vimeo'
    PLATFORM_XING = 'xing'
    PLATFORM_ANGELLIST = 'angellist'
    PLATFORM_CRUNCHBASE = 'crunchbase'
    PLATFORM_STACKEXCHANGE_NETWORK = 'stackexchange network'
    PLATFORM_WHATSAPP = 'whatsapp'
    PLATFORM_YELP = 'yelp'

    def self.all
      Platforms.constants
    end

    def self.show(const_name:)
      Platforms.const_get(const_name)
    rescue NameError
      # raise Error,
      # "#{const_name} platform don't support please read our supported platforms =>  #{Platforms.all.join(',')}"
      ''
    end
  end

  # all regex for all platforms
  class Regexes
    ANGELLIST_URL_REGEX = {
      # https://angel.co/company/twitter, https://angel.co/company/twitter/culture
      company: %r{(?:https?:)?//angel\.co/company/(?<company>[A-Za-z0-9_-]+)(?:/(?<company_subpage>[A-Za-z0-9-]+))?},
      # https://angel.co/company/twitter/jobs/576275-engineering-manager
      job: %r{
      (?:https?:)?//angel\.co/company/(?<company>[A-Za-z0-9_-]+)
       /jobs/(?<job_permalink>(?<job_id>[0-9]+)-(?<job_slug>[A-Za-z0-9-]+))
      }x,
      # https://angel.co/p/naval, https://angel.co/u/karllorey
      user: %r{(?:https?:)?//angel\.co/(?<type>u|p)/(?<user>[A-Za-z0-9_-]+)}
    }.freeze

    CRUNCHBASE_URL_REGEX = {
      # http://crunchbase.com/organization/acme-corp
      company: %r{(?:https?:)?//crunchbase\.com/organization/(?<organization>[A-Za-z0-9_-]+)},
      # http://crunchbase.com/person/karl-lorey
      person: %r{(?:https?:)?//crunchbase\.com/person/(?<person>[A-Za-z0-9_-]+)}
    }.freeze

    EMAIL_URL_REGEX = {
      # jeff@amazon.com, mailto:jeff@amazon.com, mailto:plususer+test@gmail.com
      email: /(?:mailto:)?(?<email>[A-Za-z0-9_.+-]+@[A-Za-z0-9_.-]+\.[A-Za-z]+)/,
      email_without_extract: /\A[\w+\-.]+@[a-z\d-]+(\.[a-z\d-]+)*\.[a-z]+\z/i
    }.freeze

    FACEBOOK_URL_REGEX = {
      # http://fb.com/peter_parker-miller, https://facebook.com/peter.parker, https://facebook.com/peterparker
      username: %r{
      (?:https?:)?//(?:www\.)?(?:facebook|fb)\.com/(?<profile>(?![A-Za-z]+\.php)
      (?!marketplace|gaming|watch|me|messages|help|search|groups)[A-Za-z0-9_\-.]+)/?
      }x,
      # https://www.facebook.com/100004123456789, https://www.facebook.com/profile.php?id=100004123456789
      profile_id: %r{(?:https?:)?//(?:www\.)?facebook\.com/(?:profile\.php\?id=)?(?<id>[0-9]+)}
    }.freeze

    GITHUB_URL_REGEX = {
      # https://github.com/talaatmagdyx/socials_regex
      repo: %r{(?:https?:)?//(?:www\.)?github\.com/(?<login>[A-Za-z0-9_-]+)/(?<repo>[A-Za-z0-9_-]+)/?},
      # https://github.com/talaatmagdyx
      user: %r{(?:https?:)?//(?:www\.)?github\.com/(?<login>[A-Za-z0-9_-]+)/?}
    }.freeze

    HACKERNEWS_URL_REGEX = {
      # An item can be a post or a direct link to a comment.
      # https://news.ycombinator.com/item?id=23290375
      item: %r{(?:https?:)?//news\.ycombinator\.com/item\?id=(?<item>[0-9]+)},
      # https://news.ycombinator.com/user?id=CamelCaps
      user: %r{(?:https?:)?//news\.ycombinator\.com/user\?id=(?<user>[A-Za-z0-9_-]+)}
    }.freeze

    INSTAGRAM_URL_REGEX = {
      # The rules:
      # Matches with one . in them disco.dude but not two .. disco..dude
      # Ending period not matched discodude.
      # Match underscores _disco__dude
      # Max characters of 30 1234567890123456789012345678901234567890
      # https://instagram.com/__disco__dude
      profile: %r{(?:https?:)?//(?:www\.)?(?:instagram\.com|instagr\.am)/
       (?<username>[A-Za-z0-9_](?:(?:[A-Za-z0-9_]|(?:\.(?!\.))){0,28}(?:[A-Za-z0-9_]))?)}x
    }.freeze

    LINKEDIN_URL_REGEX = {
      # https://fr.linkedin.com/school/université-grenoble-alpes/
      company: %r{(?:https?:)?//(?:\w+\.)?linkedin\.com/(?<company_type>
      (?:company|school))/(?<company_permalink>[A-Za-z0-9\-À-ÿ.]+)/?}x,
      # https://www.linkedin.com/feed/update/urn:li:activity:6665508550111912345/
      post: %r{(?:https?:)?//(?:\w+\.)?linkedin\.com/feed/update/urn:li:activity:(?<activity_id>[0-9]+)/?},
      # https://www.linkedin.com/in/talaatmagdyx/
      profile: %r{(?:https?:)?//(?:\w+\.)?linkedin\.com/in/(?<permalink>[\w\-_À-ÿ%]+)/?}
    }.freeze

    MEDIUM_URL_REGEX = {
      # https://medium.com/does-exist/some-post-123abc
      post: %r{(?:https?:)?//medium\.com/(?:(?:@(?<username>[A-Za-z0-9]+))|(?<publication>[a-z-]+))/
(?<slug>[a-z0-9-]+)-(?<post_id>[A-Za-z0-9]+)(?:\?.*)?}x,
      # https://onezero.medium.com/what-facebooks-remote-work-policy-means-for-the-future-of-tech-salaries-everywhere-edf859226b62?source=grid_home------
      # Can't match these with the regular post regex as redefinitions of subgroups are not allowed in pythons regex.
      post_of_subdomain_publication: %r{(?:https?:)?//(?<publication>(?!www)[a-z-]+)\.medium\.com/
(?<slug>[a-z0-9-]+)-(?<post_id>[A-Za-z0-9]+)(?:\?.*)?}x,
      # https://medium.com/@karllorey
      user: %r{(?:https?:)?//medium\.com/@(?<username>[A-Za-z0-9]+)(?:\?.*)?},
      # Now redirects to new user profiles. Follow with a head or get request.
      # https://medium.com/u/b3d3d3653c2c?source=post_page-----da92b81b85ef----------------------
      user_by_id: %r{(?:https?:)?//medium\.com/u/(?<user_id>[A-Za-z0-9]+)(?:\?.*)}
    }.freeze

    PHONE_URL_REGEX = {
      # Should be cleaned afterwards to strip dots, spaces, etc.
      # tel:+49 900 123456
      # tel:+49900123456
      number: /(?:tel|phone|mobile):(?<number>\+?[0-9. -]+)/
    }.freeze

    REDDIT_URL_REGEX = {
      # https://old.reddit.com/user/ar-guetita
      # https://reddit.com/u/ar-guetita
      user: %r{(?:https?:)?//(?:[a-z]+\.)?reddit\.com/(?:u(?:ser)?)/(?<username>[A-Za-z0-9\-_]*)/?}
    }.freeze

    SKYPE_URL_REGEX = {
      # Matches Skype's URLs to add contact, call, chat. More info at Skype SDK's docs.
      # Examples:
      # skype:echo123
      # skype:echo123?call
      profile: /(?:(?:callto|skype):)(?<username>[a-z][a-z0-9.,\-_]{5,31})(?:\?(?:add|call|chat|sendfile|userinfo))?/
    }.freeze

    SNAPCHAT_URL_REGEX = {
      # https://www.snapchat.com/add/example_user/
      profile: %r{(?:https?:)?//(?:www\.)?snapchat\.com/add/(?<username>[A-Za-z0-9._-]+)/?}
    }.freeze

    STACKEXCHANGE_URL_REGEX = {
      # This is the meta-platform above stackoverflow, etc. Username can be changed at any time, user_id is persistent.
      # https://www.stackexchange.com/users/12345/example_user/
      user: %r{(?:https?:)?//(?:www\.)?stackexchange\.com/users/(?<id>[0-9]+)/(?<username>[A-Za-z0-9\-_.]+)/?}
    }.freeze

    STACKEXCHANGE_NETWORK_URL_REGEX = {
      # While there are some "named" communities in the stackexchange network like stackoverflow,
      # many only exist as subdomains, i.e. gaming.stackexchange.com.
      # Again, username can be changed at any time, user_id is persistent.
      # https://gaming.stackexchange.com/users/304007/talaat-magdy
      user: %r{(?:https?:)?//(?:(?<community>[a-z]+(?!www))\.)?stackexchange\.com/users/
(?<id>[0-9]+)/(?<username>[A-Za-z0-9\-_.]+)/?}x
    }.freeze

    STACKOVERFLOW_URL_REGEX = {
      # https://stackoverflow.com/questions/12345/how-to-embed
      question: %r{(?:https?:)?//(?:www\.)?stackoverflow\.com/questions/(?<id>[0-9]+)/(?<title>[A-Za-z0-9\-_.]+)/?},
      # Username can be changed at any time, user_id is persistent.
      # https://stackoverflow.com/users/13916928/talaat-magdy
      user: %r{(?:https?:)?//(?:www\.)?stackoverflow\.com/users/(?<id>[0-9]+)/(?<username>[A-Za-z0-9\-_.]+)/?}
    }.freeze

    TELEGRAM_URL_REGEX = {
      # Matches for t.me, telegram.me and telegram.org.
      # Examples:
      # https://t.me/example_username/
      profile: %r{(?:https?:)?//(?:t(?:elegram)?\.me|telegram\.org)/(?<username>[a-z0-9_]{5,32})/?}
    }.freeze

    TWITTER_URL_REGEX = {
      # https://twitter.com/karllorey/status/1259924082067374088
      status: %r{(?:https?:)?//(?:[A-Za-z]+\.)?twitter\.com/@?(?<username>[A-Za-z0-9_]+)/status/(?<tweet_id>[0-9]+)/?},
      # Allowed for usernames are alphanumeric characters and underscores.
      # http://twitter.com/@talaatmagdyx
      # http://twitter.com/talaatmagdyx
      # https://twitter.com/talaatmagdyx
      user: %r{(?:https?:)?//(?:[A-Za-z]+\.)?twitter\.com/@?(?!home|share|privacy|tos)(?<username>[A-Za-z0-9_]+)/?}
    }.freeze

    VIMEO_URL_REGEX = {
      # https://vimeo.com/user12345
      user: %r{(?:https?:)?//vimeo\.com/user(?<id>[0-9]+)},
      # https://vimeo.com/123456789
      # https://player.vimeo.com/video/148751763
      video: %r{(?:https?:)?//(?:(?:www\.)?vimeo\.com|player\.vimeo\.com/video)/(?<id>[0-9]+)}
    }.freeze

    XING_URL_REGEX = {
      # Default slugs are Firstname_Lastname. If several people with the same name exist, a number is appended.
      # https://www.xing.com/profile/Tobias_Zilbersahn5
      profile: %r{(?:https?:)?//(?:www\.)?xing\.com/profile/(?<slug>[A-z0-9\-_]+)}
    }.freeze

    YOUTUBE_URL_REGEX = {
      # https://www.youtube.com/channel/UCxyz123456789
      channel: %r{(?:https?:)?//(?:[A-z]+\.)?youtube\.com/channel/(?<id>[A-z0-9\-_]+)},
      # https://www.youtube.com/user/username123
      user: %r{(?:https?:)?//(?:[A-z]+\.)?youtube\.com/user/(?<username>[A-z0-9]+)},
      # https://www.youtube.com/watch?v=dQw4w9WgXcQ
      # https://www.youtube.com/watch?v=dQw4w9WgXcQ
      # https://www.youtube.com/embed/dQw4w9WgXcQ
      # https://www.youtube.com/watch?v=6_b7RDuLwcI
      video: %r{(?:https?:)?//(?:(?:www\.)?youtube\.com/(?:watch\?v=|embed/)|youtu\.be/)(?<id>[A-z0-9\-_]+)}
    }.freeze

    WHATSAPP_URL_REGEX = {
      # https://wa.me/1234567890
      phone: %r{(?:https?:)?//(?:wa\.me/)(?<number>\+?[0-9. -]+)}
    }.freeze

    YELP_URL_REGEX = {
      # https://www.yelp.com/biz/example-business
      company: %r{(?:https?://)?(?:www\.)?yelp\.com/biz/(?<company>[A-Za-z0-9_-]+)}
    }.freeze

    def self.match?(input_str:, regex:)
      input_str.match(regex) ? true : false
    end

    def self.all
      Regexes.constants
    end
  end
end
